src/server/pfs/s3/s3_test.go

```diff
@@ -10,50 +10,39 @@ import (
 	"fmt"
 	"io"
 	"io/ioutil"
-	"net/http"
 	"os"
 	"path/filepath"
 	"sort"
 	"strings"
+	"sync"
 	"testing"
 	"time"
 
 	minio "github.com/minio/minio-go"
 	"github.com/pachyderm/pachyderm/src/client"
 	"github.com/pachyderm/pachyderm/src/client/pkg/require"
-	"github.com/pachyderm/pachyderm/src/server/pfs/server"
-	"github.com/pachyderm/pachyderm/src/server/pkg/backoff"
 	tu "github.com/pachyderm/pachyderm/src/server/pkg/testutil"
 )
 
-func serve(t *testing.T) (*http.Server, *client.APIClient, *minio.Client) {
-	t.Helper()
+var (
+	pachClient *client.APIClient
```

Comment:
I think `pachClient` and `minioClient` should be scoped into `func clients` - there's no reason for them to be global.

Reply:
We seem to rely on singleton clients for integration tests everywhere else. I'm not sure why, exactly. @msteffen would you maybe know?

Reply:
Basically because GRPC discourages you from calling `grpc.Dial` more than once[1]. We have a library called `ServiceEnv` that manages stubs for you—we still need to migrate some of our tests to use it, but this could be a good opportunity to start? You'd have to break `clients()` into `getPachClient()` and `getMinioClient()`, or add minio to ServiceEnv, but it might clean up this code the way you like.

[1] via https://groups.google.com/forum/?pli=1#!searchin/grpc-io/clientconn%7Csort:date/grpc-io/bh-sKYeOEr0/V2lh6O2yCAAJ -- "you should not try creating the *grpc.ClientConn for each request. Instead create it once and cache it. You can create the stub just once and cache it (they are safe to use concurrently form multiple goroutines). But that is not necessary; you could also create the stub for each request, using the cached connection."